pkg/server/handler/v1alpha1/workflowrun.go

```diff
@@ -459,3 +466,150 @@ func GetContainerLogs(ctx context.Context, project, workflow, workflowrun, tenan
 
 	return folderReader, headers, nil
 }
+
+// ReceiveStageArtifacts receives artifacts produced by workflowrun stage.
+func ReceiveStageArtifacts(ctx context.Context, workflowrun, namespace, stage string) error {
+	// get workflowrun
+	wfr, err := handler.K8sClient.CycloneV1alpha1().WorkflowRuns(namespace).Get(workflowrun, metav1.GetOptions{})
+	if err != nil {
+		log.Errorf("get wfr %s/%s error %s", namespace, workflowrun, err)
+		return err
+	}
+
+	// get tenant, project, workflow from workflowrun
+	tenant := common.NamespaceTenant(namespace)
+	var project, workflow string
+	if wfr.Labels != nil {
+		project = wfr.Labels[meta.LabelProjectName]
+		workflow = wfr.Labels[meta.LabelWorkflowName]
+	}
+	if project == "" || workflow == "" {
+		return fmt.Errorf("failed to get project or workflow from workflowrun labels")
+	}
+
+	request := contextutil.GetHTTPRequest(ctx)
+
+	file, fileHeader, err := request.FormFile(stageArtifactFormFileKey)
+	if err != nil {
+		log.Infof("Form file by key %s error: %v", stageArtifactFormFileKey, err)
+		return err
+	}
+
+	defer func() {
+		if err := file.Close(); err != nil {
+			log.Errorf("Fail to close file as: %v", err)
+		}
+	}()
+
+	folder, err := getArtifactFolder(tenant, project, workflow, workflowrun, stage)
+	if err != nil {
+		log.Errorf("Get workflowrun %s artifact folder failed: %v", workflowrun, err)
+		return err
+	}
+
+	// create artifact folders.
+	fileutil.CreateDirectory(folder)
+	// copy file
+	f, err := os.OpenFile(fmt.Sprintf("%s/%s", folder, fileHeader.Filename), os.O_WRONLY|os.O_CREATE, 0666)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if err := f.Close(); err != nil {
+			log.Errorf("Fail to close file as: %v", err)
+		}
+	}()
+
+	_, err = io.Copy(f, file)
+	if err != nil {
+		log.Infof("copy artifact %s error: %v", fileHeader.Filename, err)
+	}
+
+	return err
+}
+
+// ListArtifacts handles the request to list artifacts produced in a workflowRun.
+func ListArtifacts(ctx context.Context, project, workflow, workflowrun, tenant string) (*types.ListResponse, error) {
+	var artifacts []api.StageArtifact
+
+	wf, err := handler.K8sClient.CycloneV1alpha1().Workflows(common.TenantNamespace(tenant)).Get(workflow, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, stage := range wf.Spec.Stages {
+		artifactFolder, _ := getArtifactFolder(tenant, project, workflow, workflowrun, stage.Name)
+		artifactFolderInfo, err := os.Stat(artifactFolder)
+		if os.IsNotExist(err) {
+			continue
+		}
+		if !artifactFolderInfo.IsDir() {
+			continue
+		}
+
+		files, err := ioutil.ReadDir(artifactFolder)
+		if err != nil {
+			log.Infof("Readdir %s error:%v", artifactFolder, err)
+			return nil, err
+		}
+
+		for _, file := range files {
+			// Only support display files
+			if file.IsDir() {
```

Comment:
It will be a little confused if we collect folders but not display them.

Reply:
we can not collect folders, all files will be compressed to a tar file.